### PR TITLE
Propagate labels through various Structure operations

### PR DESCRIPTION
**Pull request by stefsmeets**
## Summary

This PR follows up from #3137 and makes sure that labels are propagated correctly throughout structure, for example, when making supercells, inserting/deleting sites, applying translations/rotations, etc.

This PR also fixes a bug where the labels were incorrectly assigned when the CIF contains multiple species with different names.

Major changes:

- feature 1: Propagate labels in operations on structure
- fix 1: Correctly generate labels when cif contains multiple species with same name

## Checklist

- [x] Google format doc strings added. Check with `ruff`.
- [x] Type annotations included. Check with `mypy`.
- [x] Tests added for new features/fixes.
- [x] If applicable, new classes/functions/modules have [`duecredit`](https://github.com/duecredit/duecredit) `@due.dcite` decorators to reference relevant papers by DOI ([example](https://github.com/materialsproject/pymatgen/blob/91dbe6ee9ed01d781a9388bf147648e20c6d58e0/pymatgen/core/lattice.py#L1168-L1172))

Tip: Install `pre-commit` hooks to auto-check types and linting before every commit:

```sh
pip install -U pre-commit
pre-commit install
```

### CODE REVIEW

**Comment by janosh** [comment]
Thank you very much @stefsmeets, this is great! 🙏 

Just to get an overview, are there more structure operations that are known to drop the site labels? Supercells, inserting/deleting sites, applying translations/rotations are all very important. Just want to leave a record here of what else might need doing in the future.

**Comment by stefsmeets** [comment]
@janosh Good point, I think these functions could still be updated:

- [ ] `IStructure.from_spacegroup`
- [ ] `IStructure.from_magnetic_spacegroup`
- [ ] `IMolecule.get_boxed_structure`
- [ ] `IMolecule.get_centered_structure`
- [ ] `IStructure.interpolate`
- [ ] Everything to do with `Neighbor`/`PeriodicNeighbor`

I will update the PR to update these today or tomorrow.

There is also the CIF writer which drops the labels. I looked at this before, but I found it a bit tricky to update.

**Comment by janosh** [comment]
> There is also the CIF writer which drops the labels. I looked at this before, but I found it a bit tricky to update.

Feel free to just leave a todo comment for CIFParser so that we know it should still be done at some point.

**Comment by stefsmeets** [comment]
I had a go at updating the CifWriter with labels. Since the label defaults to `Site.species_string`, it turns out that it's a little bit more straightforward than when I first looked at this. 

I also updated `site.label` to be a property. This helps generate the appropriate label for methods that update `Site` inplace (such as `SitesCollection.add_oxidation_state_by_*`).

Let me know what you think.

File: pymatgen/io/cif.py [hunk 1]
```diff
@@ -1022,9 +1038,13 @@ def get_matching_coord(coord):
...
                 if self.feature_flags["magcif"]:
-                    coords, magmoms = self._unique_coords(tmp_coords, magmoms_in=tmp_magmom, lattice=lattice)
+                    coords, magmoms, new_labels = self._unique_coords(
+                        tmp_coords, magmoms=tmp_magmom, labels=labels, lattice=lattice
+                    )
                 else:
-                    coords, magmoms = self._unique_coords(tmp_coords)
+                    coords, magmoms, new_labels = self._unique_coords(tmp_coords, labels=labels)
+
+                assert len(new_labels) == len(coords)
```
[the remark this answers is not in the document]
Comment (stefsmeets) [reply pymatgen/io/cif.py:1047]: Yeah, agreed. I was using this while debugging. This really should not happen though, so I will remove the assert altogether.

**Review by janosh — APPROVED** [review]
@stefsmeets Thanks a lot for this PR! You really went the extra mile.

**Comment by mkhorton** [comment]
I noticed that performing a `Structure.replace_species` operation leaves the label intact, leading to a misleading label or a confusing end-user experience (for example, replacing `Au` with `{"Au": 0.5, "Cu": 0.5}` will leave the label as `Au`). Do you have any thoughts on this @stefsmeets?

**Comment by stefsmeets** [comment]
> I noticed that performing a `Structure.replace_species` operation leaves the label intact, leading to a misleading label or a confusing end-user experience (for example, replacing `Au` with `{"Au": 0.5, "Cu": 0.5}` will leave the label as `Au`). Do you have any thoughts on this @stefsmeets?

I don't have any particular thoughts on this, but I can see how it would be confusing. What would you expect to happen?

**Comment by stefsmeets** [comment]
I can have a go at this next week. I created a new issue out of it to keep track.

**Comment by fxcoudert** [comment]
I think this caused a regression in CIF output: https://github.com/materialsproject/pymatgen/issues/3761

**Pull request merged**